File: internal/swarm/pathfind/navmesh/abstractfile.go
// SPDX-FileCopyrightText: 2026 Melg Eight <public.melg8@gmail.com>
//
// SPDX-License-Identifier: MIT

package navmesh

import (
    "encoding/binary"
    "fmt"
    "math"
    "time"
)

// The abstract sidecar format: the persistent coarse layer of one
// region (docs/fastpath_research.md section 2). The pack build
// serializes the cluster graph next to the tile (the X_Y.ab
// sidecar); the runtime loads it instead of re-scanning the decoded
// tile, so the coarse search starts warm and the tile decodes stay
// inside the refinement hops.
//
// The wire: a 56 byte header, the node records (the inline edge
// index lists), the edge records and the run length encoded link
// components. The derived structures (the cluster index map and the
// per node class dedupe) rebuild at the decode - the edge dedupe of
// the build keeps one edge per class, so the first class mapping per
// stored edge reconstructs the map exactly.
//
// Version 2 adds the tile checksum guard (the head and the tail
// CRC32 of the tile file bytes in the header bytes 48..56): the
// deployment copies that lose the file mtimes (a plain cp or an
// archive extraction) keep serving the sidecars, the mtime only
// packs stay on the version 1 guard.

const (
    abstractMagic        = 0x31424153 // 'SAB1' little endian
    abstractVersion      = 2
    abstractVersionV1    = 1 // the mtime only guard (the legacy sidecars)
    abstractHeaderSize   = 56
    abstractEdgeWireSize = 64
)

// EncodeAbstract serializes a region cluster graph.
func EncodeAbstract(abstract *regionAbstract) ([]byte, error) {
    if abstract == nil || len(abstract.nodes) == 0 {
        return nil, fmt.Errorf("%w: the empty abstract", ErrBadTile)
    }
    nodeBytes := 0
    for i := range abstract.nodes {
        nodeBytes += 8 + len(abstract.nodes[i].edges)*4
    }
    runs := compRunsOf(abstract.comps)
    size := abstractHeaderSize + nodeBytes +
        len(abstract.edges)*abstractEdgeWireSize + len(runs)*8
    data := make([]byte, size)

    put32 := func(off int, v uint32) {
        binary.LittleEndian.PutUint32(data[off:], v)
    }
    put16 := func(off int, v uint16) {
        binary.LittleEndian.PutUint16(data[off:], v)
    }
    put32(0, abstractMagic)
    if abstract.TileChecksums {
        put32(4, abstractVersion)
    } else {
        put32(4, abstractVersionV1)
    }
    put32(8, uint32(int32(abstract.key.Col)))
    put32(12, uint32(int32(abstract.key.Row)))
    put32(16, uint32(abstract.polys))
    put32(20, uint32(len(abstract.nodes)))
    put32(24, uint32(len(abstract.edges)))
    put32(28, uint32(len(runs)))
    binary.LittleEndian.PutUint64(data[32:], uint64(abstract.TileSize))
    binary.LittleEndian.PutUint64(data[40:],
        uint64(abstract.TileModTime.UnixNano()))
    if abstract.TileChecksums {
        put32(48, abstract.TileHeadCRC)
        put32(52, abstract.TileTailCRC)
    }

    offset := abstractHeaderSize
    for i := range abstract.nodes {
        node := &abstract.nodes[i]
        data[offset] = byte(node.id)
        put16(offset+2, uint16(len(node.edges)))
        offset += 8
        for _, edgeIdx := range node.edges {
            put32(offset, uint32(edgeIdx))
            offset += 4
        }
    }
    for i := range abstract.edges {
        edge := &abstract.edges[i]
        base := offset + i*abstractEdgeWireSize
        put16(base, uint16(int16(edge.to.Col)))
        put16(base+2, uint16(int16(edge.to.Row)))
        data[base+4] = byte(edge.to.ID)
        binary.LittleEndian.PutUint64(data[base+8:], uint64(edge.toRef))
        put32(base+16, uint32(edge.link))
        binary.LittleEndian.PutUint64(data[base+20:], math.Float64bits(edge.mid.X))
        binary.LittleEndian.PutUint64(data[base+28:], math.Float64bits(edge.mid.Y))
        binary.LittleEndian.PutUint64(data[base+36:], math.Float64bits(edge.mid.Z))
        put32(base+44, edge.srcComp)
        put32(base+48, edge.dstComp)
        put16(base+52, uint16(int16(edge.extCol)))
        put16(base+54, uint16(int16(edge.extRow)))
        put32(base+56, edge.extPoly)
        data[base+60] = edge.srcArea
        data[base+61] = edge.dstArea
    }
    offset += len(abstract.edges) * abstractEdgeWireSize
    for _, run := range runs {
        put32(offset, run.value)
        put32(offset+4, run.count)
        offset += 8
    }

    return data, nil
}

// DecodeAbstract parses one abstract sidecar.
func DecodeAbstract(data []byte) (*regionAbstract, error) {
    if len(data) < abstractHeaderSize {
        return nil, fmt.Errorf("%w: the abstract %d bytes is too short",
            ErrBadTile, len(data))
    }
    if magic := binary.LittleEndian.Uint32(data[0:]); magic != abstractMagic {
        return nil, fmt.Errorf("%w: the abstract magic 0x%x", ErrBadTile,
            magic)
    }
    if version := binary.LittleEndian.Uint32(data[4:]); version !=
        abstractVersion && version != abstractVersionV1 {
        return nil, fmt.Errorf("%w: the abstract version %d", ErrBadTile,
            version)
    }
    checksums := binary.LittleEndian.Uint32(data[4:]) == abstractVersion
    u32 := func(off int) uint32 { return binary.LittleEndian.Uint32(data[off:]) }
    col := int16(u32(8))
    row := int16(u32(12))
    polys := int(u32(16))
    nodeCount := int(u32(20))
    edgeCount := int(u32(24))
    runCount := int(u32(28))
    tileSize := int64(binary.LittleEndian.Uint64(data[32:]))
    tileModTime := time.Unix(0,
        int64(binary.LittleEndian.Uint64(data[40:])))

    abstract := &regionAbstract{
        key:   RegionKey{Col: col, Row: row},
        index: make(map[clusterID]int32, nodeCount),
        edges: make([]abstractEdge, 0, edgeCount),
        comps: make([]uint32, polys),
        polys: polys,

        TileSize:      tileSize,
        TileModTime:   tileModTime,
        TileChecksums: checksums,
        TileHeadCRC:   binary.LittleEndian.Uint32(data[48:]),
        TileTailCRC:   binary.LittleEndian.Uint32(data[52:]),
    }

    offset := abstractHeaderSize
    for i := 0; i < nodeCount; i++ {
        if offset+8 > len(data) {
            return nil, fmt.Errorf("%w: the abstract nodes truncate",
                ErrBadTile)
        }
        id := clusterID(data[offset])
        edgeN := int(binary.LittleEndian.Uint16(data[offset+2:]))
        offset += 8
        if offset+edgeN*4 > len(data) {
            return nil, fmt.Errorf("%w: the abstract node edges truncate",
                ErrBadTile)
        }
        node := abstractNode{
            id:      id,
            edges:   make([]int32, 0, edgeN),
            classes: make(map[abstractClassKey]int32, edgeN),
        }
        for j := 0; j < edgeN; j++ {
            idx := int32(binary.LittleEndian.Uint32(data[offset:]))
            offset += 4
            if idx < 0 || int(idx) >= edgeCount {
                return nil, fmt.Errorf(
                    "%w: the abstract edge index %d", ErrBadTile, idx)
            }
            node.edges = append(node.edges, idx)
        }
        abstract.nodes = append(abstract.nodes, node)
        abstract.index[id] = int32(len(abstract.nodes) - 1)
    }
    if offset+edgeCount*abstractEdgeWireSize > len(data) {
        return nil, fmt.Errorf("%w: the abstract edges truncate",
            ErrBadTile)
    }
    for i := 0; i < edgeCount; i++ {
        base := offset + i*abstractEdgeWireSize
        edge := abstractEdge{
            to: clusterKey{
                Col: int16(binary.LittleEndian.Uint16(data[base:])),
                Row: int16(binary.LittleEndian.Uint16(data[base+2:])),
                ID:  clusterID(data[base+4]),
            },
            toRef: PolyRef(binary.LittleEndian.Uint64(data[base+8:])),
            link:  int32(binary.LittleEndian.Uint32(data[base+16:])),
            mid: Pos{
                X: math.Float64frombits(
                    binary.LittleEndian.Uint64(data[base+20:])),
                Y: math.Float64frombits(
                    binary.LittleEndian.Uint64(data[base+28:])),
                Z: math.Float64frombits(
                    binary.LittleEndian.Uint64(data[base+36:])),
            },
            srcComp: binary.LittleEndian.Uint32(data[base+44:]),
            dstComp: binary.LittleEndian.Uint32(data[base+48:]),
            extCol: int16(binary.LittleEndian.Uint16(
                data[base+52:])),
            extRow: int16(binary.LittleEndian.Uint16(
                data[base+54:])),
            extPoly: binary.LittleEndian.Uint32(data[base+56:]),
            srcArea: data[base+60],
            dstArea: data[base+61],
        }
        abstract.edges = append(abstract.edges, edge)
    }
    offset += edgeCount * abstractEdgeWireSize
    if offset+runCount*8 > len(data) {
        return nil, fmt.Errorf("%w: the abstract components truncate",
            ErrBadTile)
    }
    cursor := 0
    for i := 0; i < runCount; i++ {
        base := offset + i*8
        value := binary.LittleEndian.Uint32(data[base:])
        count := int(binary.LittleEndian.Uint32(data[base+4:]))
        for j := 0; j < count; j++ {
            if cursor >= polys {
                return nil, fmt.Errorf(
                    "%w: the abstract components overflow", ErrBadTile)
            }
            abstract.comps[cursor] = value
            cursor++
        }
    }

    // The class dedupe rebuilds from the stored edges: the build kept
    // one edge per class, the first mapping per stored edge
    // reconstructs the map.
    for i := range abstract.nodes {
        node := &abstract.nodes[i]
        for _, edgeIdx := range node.edges {
            edge := &abstract.edges[edgeIdx]
            node.classes[abstractClassKey{
                to:      edge.to,
                srcArea: edge.srcArea,
                dstArea: edge.dstArea,
                srcComp: edge.srcComp,
            }] = edgeIdx
        }
    }

    return abstract, nil
}

// compRun is one run of the link component array.
type compRun struct {
    value uint32
    count uint32
}

// compRunsOf folds the component array into the runs.
func compRunsOf(comps []uint32) []compRun {
    runs := make([]compRun, 0, len(comps)/8+1)
    for _, value := range comps {
        if n := len(runs); n > 0 && runs[n-1].value == value {
            runs[n-1].count++

            continue
        }
        runs = append(runs, compRun{value: value, count: 1})
    }

    return runs
}

File: pathfind-rs/src/abstractfile.rs
// SPDX-FileCopyrightText: 2026 Melg Eight <public.melg8@gmail.com>
//
// SPDX-License-Identifier: MIT

//! The abstract sidecar format: the persistent coarse layer of one
//! region. The pack build serializes the cluster graph next to the
//! tile (the X_Y.ab sidecar); the runtime loads it instead of
//! re-scanning the decoded tile.
//!
//! This port carries the runtime half (the decode); the encoder lives
//! with the Go navbuild tool that produces the sidecars.

use crate::abstract_graph::{
    AbstractClassKey, AbstractEdge, AbstractNode, ClusterId, ClusterKey, RegionAbstract,
};
use crate::tile::RegionKey;
use crate::NavError;

const ABSTRACT_MAGIC: u32 = 0x3142_4153; // 'SAB1' little endian
const ABSTRACT_VERSION: u32 = 2;
const ABSTRACT_VERSION_V1: u32 = 1;
const ABSTRACT_HEADER_SIZE: usize = 56;
const ABSTRACT_EDGE_WIRE_SIZE: usize = 64;

fn err_bad(msg: String) -> NavError {
    NavError(format!("not a navmesh tile: {msg}"))
}

fn u32_le(data: &[u8], off: usize) -> u32 {
    u32::from_le_bytes([data[off], data[off + 1], data[off + 2], data[off + 3]])
}

fn u16_le(data: &[u8], off: usize) -> u16 {
    u16::from_le_bytes([data[off], data[off + 1]])
}

fn f64_le(data: &[u8], off: usize) -> f64 {
    f64::from_le_bytes([
        data[off],
        data[off + 1],
        data[off + 2],
        data[off + 3],
        data[off + 4],
        data[off + 5],
        data[off + 6],
        data[off + 7],
    ])
}

/// Parses one abstract sidecar.
pub fn decode_abstract(data: &[u8]) -> Result<RegionAbstract, NavError> {
    if data.len() < ABSTRACT_HEADER_SIZE {
        return Err(err_bad(format!(
            "the abstract {} bytes is too short",
            data.len()
        )));
    }
    if u32_le(data, 0) != ABSTRACT_MAGIC {
        return Err(err_bad(format!("the abstract magic 0x{:x}", u32_le(data, 0))));
    }
    let version = u32_le(data, 4);
    if version != ABSTRACT_VERSION && version != ABSTRACT_VERSION_V1 {
        return Err(err_bad(format!("the abstract version {version}")));
    }
    let checksums = version == ABSTRACT_VERSION;
    let col = u32_le(data, 8) as i16;
    let row = u32_le(data, 12) as i16;
    let polys = u32_le(data, 16) as usize;
    let node_count = u32_le(data, 20) as usize;
    let edge_count = u32_le(data, 24) as usize;
    let run_count = u32_le(data, 28) as usize;
    let tile_size = u64::from_le_bytes(data[32..40].try_into().unwrap()) as i64;
    let tile_mod_time_unix_nano =
        u64::from_le_bytes(data[40..48].try_into().unwrap()) as i64;

    let mut abstract_ = RegionAbstract {
        key: RegionKey { col, row },
        index: std::collections::HashMap::with_capacity(node_count),
        edges: Vec::with_capacity(edge_count),
        comps: vec![0u32; polys],
        nodes: Vec::new(),
        polys,
        tile_size,
        tile_mod_time_unix_nano,
        tile_checksums: checksums,
        tile_head_crc: u32_le(data, 48),
        tile_tail_crc: u32_le(data, 52),
    };

    let mut offset = ABSTRACT_HEADER_SIZE;
    for _ in 0..node_count {
        if offset + 8 > data.len() {
            return Err(err_bad("the abstract nodes truncate".to_string()));
        }
        let id: ClusterId = data[offset];
        let edge_n = u16_le(data, offset + 2) as usize;
        offset += 8;
        if offset + edge_n * 4 > data.len() {
            return Err(err_bad("the abstract node edges truncate".to_string()));
        }
        let mut node = AbstractNode {
            id,
            edges: Vec::with_capacity(edge_n),
            classes: std::collections::HashMap::with_capacity(edge_n),
        };
        for _ in 0..edge_n {
            let idx = u32_le(data, offset) as i32;
            offset += 4;
            if idx < 0 || (idx as usize) >= edge_count {
                return Err(err_bad(format!("the abstract edge index {idx}")));
            }
            node.edges.push(idx);
        }
        abstract_.index.insert(id, abstract_.nodes.len() as i32);
        abstract_.nodes.push(node);
    }
    if offset + edge_count * ABSTRACT_EDGE_WIRE_SIZE > data.len() {
        return Err(err_bad("the abstract edges truncate".to_string()));
    }
    for i in 0..edge_count {
        let base = offset + i * ABSTRACT_EDGE_WIRE_SIZE;
        let edge = AbstractEdge {
            to: ClusterKey {
                col: u16_le(data, base) as i16,
                row: u16_le(data, base + 2) as i16,
                id: data[base + 4],
            },
            to_ref: u64::from_le_bytes(data[base + 8..base + 16].try_into().unwrap()),
            link: u32_le(data, base + 16) as i32,
            mid: crate::query::Pos {
                x: f64_le(data, base + 20),
                y: f64_le(data, base + 28),
                z: f64_le(data, base + 36),
            },
            src_comp: u32_le(data, base + 44),
            dst_comp: u32_le(data, base + 48),
            ext_col: u16_le(data, base + 52) as i16,
            ext_row: u16_le(data, base + 54) as i16,
            ext_poly: u32_le(data, base + 56),
            src_area: data[base + 60],
            dst_area: data[base + 61],
        };
        abstract_.edges.push(edge);
    }
    offset += edge_count * ABSTRACT_EDGE_WIRE_SIZE;
    if offset + run_count * 8 > data.len() {
        return Err(err_bad("the abstract components truncate".to_string()));
    }
    let mut cursor = 0usize;
    for i in 0..run_count {
        let base = offset + i * 8;
        let value = u32_le(data, base);
        let count = u32_le(data, base + 4) as usize;
        for _ in 0..count {
            if cursor >= polys {
                return Err(err_bad("the abstract components overflow".to_string()));
            }
            abstract_.comps[cursor] = value;
            cursor += 1;
        }
    }

    // The class dedupe rebuilds from the stored edges: the build kept
    // one edge per class, the first mapping per stored edge
    // reconstructs the map.
    for node in abstract_.nodes.iter_mut() {
        for &edge_idx in &node.edges {
            let edge = &abstract_.edges[edge_idx as usize];
            let key = AbstractClassKey {
                to: edge.to,
                src_area: edge.src_area,
                dst_area: edge.dst_area,
                src_comp: edge.src_comp,
            };
            node.classes.insert(key, edge_idx);
        }
    }

    Ok(abstract_)
}
